### ISSUE TITLE
Short-term roadmap

### ISSUE THREAD

**Issue by latkin** [Microsoft/visualfsharp#563]
## Short-term roadmap

Now that F# 4.0 is out the door, we wanted to outline plans for Visual F# spanning the coming weeks and months.  Below represents a short-term roadmap, covering work that is in-progress or will happen in the coming weeks.  @KevinRansom is putting together a longer-term roadmap and will post that when it's ready.  We are eager to get regular code flow moving again!

#### CoreCLR

As you can see in [#499](https://github.com/Microsoft/visualfsharp/pull/499), we are working on standing up the F# compiler and runtime on CoreCLR.  When CoreCLR hits 1.0, it's important that F# is ready and supported.  In particular, getting the core F# compilation toolchain working on CoreCLR for cross-platform development is something we really want to invest in. We consider this to be one of, if not the, top priority item for us in the coming months.

We are starting with just the basics -
- Compiler
- Runtime
- Console apps

Then we will add in more and more as the platform stabilizes and we can build it -
- Interactive
- Type providers - binary compatible? Need re-compile?
- Cross-platform IDE tools based on .NET Core/CoreCLR

#### Clear out backlog of PRs

We put a lot of PRs on hold while F# 4.0 was locking down. We even have a few lingering PRs [on our old CodePlex page](http://visualfsharp.codeplex.com/sourcecontrol/list/contributions)! We are now looking forward to clearing the backlog and making room for new contributions.

In particular, there are a number of pending "cleanup" PRs that do large-scale re-arrangement and cleanup of the codebase. It will be disruptive, but we think it's best to merge these **first**, getting the hard part out of the way. Once this is done we will clear the other bugfix and performance PRs.  This will likely require PR authors to assist with resolving merge conflicts, so please bear with us.

#### Branches

Here are the branches we want to use for this upcoming work.

Branch | Purpose | Flow
-----  | ------- | ----
`master` | VS vNext/Dev15 work | Send all general PRs here
`OOB`    | VS 2015 Update/OOB update work | Cherry-pick fixes from `master`
`coreclr` | CoreCLR work | Takes merges from `OOB`

We will use git tags to mark the final commits of various releases.

In picture form:

![image](https://cloud.githubusercontent.com/assets/5943573/8992030/1fbfd676-36b1-11e5-8885-9c0a95435899.png)

The existing `fsharp4` branch will turn into `OOB` and we will delete `fsharp4`.  Once the PR backlog has been cleared, we will merge everything back into `master`, which will be where most contributions occur in the near future. `coreclr` will track with `OOB`.

With F# 4.0 development we had a confusing split where we asked some PRs to be made against `master` and others against `fsharp4`. This time it should be simpler - send all PRs to `master`, and we will simply cherry-pick the appropriate changes into the `OOB` branch for the next update release.

`master` develepment will ultimately turn into "F# 4.1" or whatever ships in the next major Visual Studio release.

`OOB` development will ultimately become the next update release. This might be timed with VS Update, or might be totally out-of-band.

**Comment by dungpa** [comment]
+1 for merging cleanup PRs first. It's a huge win in the long term.

**Comment by ovatsus** [comment]
Thanks for putting this out on the open. As part of the coreCLR work, will there be support for creating 100% F# UWP apps and ASPvNext, which I beleive are "profiles" of coreCLR without needing to have a C# entrypoint?

**Comment by latkin** [comment]
@ovatsus There are a lot of moving parts, which are easy to confuse (pretty sure I'm still confused).

The F# compiler can already output portable libraries which in theory should be consumable and runnable by ASP vNext.  But if we want support for on-the-fly compilation within ASP nNext, the compiler itself needs to be runnable on .NET Core (that's what the current work is).  Even once that's possible, there is still nontrivial plumbing to be done within ASP.NET in order for pure-F# entrypoints to be possible.

For UWP, the compiler doesn't need to run on .NET Core, it just needs to output compatible assemblies. Technically we can do that today via portable libraries. However .NET Native is a required part of the UWP pipeline, and as it stands .NET Native can't handle F# assemblies. We are working on addressing that.  But once again, there are more steps before pure-F# entrypoints are possible.

**Comment by ovatsus** [comment]
Oh, I didn't realize this was about getting the compiler itself to run on CoreCLR. What I'm mostly interested in is on having a CoreCLR FSharp.Core and having the F# compiler target them, so I can have a full UWP F# app without needing the PCLs. The problem with the PCLs is that you cannot access any of the platform specific APIs from F#, forcing you to use C# for a big chunk of the functionality.
I didn't know UWP required .NET Native, thought that was an optional optimization. That alone will make sure we won't see UWP F# apps anytime soon :(

**Comment by manofstick** [comment]
I'm happy with whatever path you choose. Just tell me what to you need me to do (happy to collapse changed into single check-in, probably not interested in my rambling paths to solutions anyway!)

**Comment by forki** [comment]
I know it's really important for some people that we follow all the platform trends and all the super innovative ASP.Net vnext project ideas, but what about real language features? Or things that improve the daily life of a F# developer.

**Comment by dsyme** [comment]
@forki   The above is the near-term roadmap for the Visual F# Tools team, specifically w.r.t. branches.  Tooling, cross-platform reach and .NET Core are, in the long run, very important to Visual F#'s viability and I believe the above are the right thing for the Visual F# Tools team to be focusing on.  

The F# language itself is in good shape and the vast majority of requests/dissatisfaction/needs we see are around tooling, integration into broader programming scenarios (cross-plat, mobile, UWP, cloud etc.). The Visual F# Tools team are always open to PRs in areas such as bug-fixing, performance, cleanup, tooling etc..

At this moment I think its best that improvements to the F# language be a separate track of work on a different delivery timeline. Currently, given the above branches, I'd recommend submitting PRs to OOB (or master) for later integration without there being a specific branch for integrating those as yet.  Right now, there's no specific timeline for F# 4.1 or 5.0 as a language design. To be honest I'm happy to let things settle for a while after getting F# 4.0 out the door. We are still, for example, in the process of seeing F# 4.0 out into the wild on Linux, Xamarin, FSharp.Compiler.Service etc.

For the language design process, my personal approach is to devote part of my time to that area (but not all).  My personal current work in this area has been to go through the http://fslang.uservoice.com suggestions to refine them to a more meaningful set of propositions. I've been marking some minor things as [approved/planned](http://fslang.uservoice.com/forums/245727-f-language/status/1225914). Certainly people can get started on those and submit them as independent PRs. However I can't give a specific timeline for how language improvements would be integrated into the repo branches outlined above.

**Comment by forki** [comment]
thanks @dsyme. I know that the language design side takes time, but it would still be interesting to see what the next steps could be. But you are right that's probably not "short term".

The second thing that I meant is tooling. Now that F# 4.0 is released, I wonder what the direction is? I know there were discussions about omnisharp, atom, VFPT, FCS and so on. There is not much about this in this short term agenda. I hope to learn more about the plans in @KevinRansom's longer-term roadmap.

**Comment by latkin** [comment]
@ovatsus Yes, the current work is to get the compiler itself (and ultimately interactive, IDE tooling, etc) to target .NET Core, so that it can run on CoreCLR, and thus cross-platform and (in theory) within ASP vNext.

Accessing platform APIs in modern Windows apps in orthogonal to CoreCLR/.NET Core. That requires *WinRT* support. Supporting WinRT is yet another big ball of wax and at this point we aren't looking to prioritize any work in that area.

Like ASP, getting support for a *pure-F# entrypoint* in UWP (assuming current .NET Native blockers are resolved) is additional plumbing and tooling, it's not just a matter of targeting the right platform.

**Comment by 7sharp9** [comment]
@latkin Does any of the UWP include shared projects for F#?

Having to rely on linked files or duplicated code is less than ideal for mobile applications.

**Comment by latkin** [comment]
@7sharp9 expanding the IDE support to include shared projects is certainly on the table, but not currently considered a top priority.  We would of welcome a PR that implemented this, though.

**Comment by forki** [comment]
@latkin are you going to change the default branch to OOB or where do you expect most PR atm?

**Comment by latkin** [comment]
I was looking at that last week, but GitHub warned be that it could have an impact on pending PRs.  I didn't want to disrupt any of those.  Do you happen to know if changing the default branch is safe in this regard?

Once I get through enough backlog I'll merge OOB into master, then everything should go against master.

**Comment by forki** [comment]
yes using NAV would even be better

2015-08-11 18:18 GMT+02:00 Lincoln Atkinson <notifications@github.com>:

**Comment by forki** [comment]
>  Do you happen to know if changing the default branch is safe in this regard?

I can't remember any issues, but who knows.

**Comment by dsyme** [comment]
I don't think changing the default branch has any negative impact, it just means GitHub will be set up to direct new PRs to that branch by default.

**Comment by latkin** [comment]
Ok I went ahead and changed the default to OOB

**Comment by Alxandr** [comment]
> The F# compiler can already output portable libraries which in theory should be consumable and runnable by ASP vNext. But if we want support for on-the-fly compilation within ASP vNext, the compiler itself needs to be runnable on .NET Core (that's what the current work is). Even once that's possible, there is still nontrivial plumbing to be done within ASP.NET in order for pure-F# entrypoints to be possible.

This plumbing is trivial (unless you include tooling, then I have no clue), and I've already done it more than once. To get F# support in DNX **today** (no CoreCLR though) see https://github.com/YoloDev/YoloDev.Dnx.FSharp.

**Comment by dsyme** [comment]
closing old discussion
